hecate_llvm_gen/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+        #[allow(unused_unsafe)]
+        unsafe { std::ffi::CStr::from_ptr(concat!($s, "\0").as_ptr() as *const i8) }
+    );
+    ($s:expr) => (
+        #[allow(unused_unsafe)]
+        unsafe { std::ffi::CStr::from_ptr(($s.to_string() + "\0").as_ptr() as *const i8) }
+    );
+}
+
+macro_rules! c_str_ptr {
+    () => (
+        c_str!().as_ptr()
+    );
+    ($s:literal) => (
+        c_str!($s).as_ptr()
+    );
+    ($s:expr) => (
+        c_str!($s).as_ptr()
+    );
+}
+
+#[repr(u32)]
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum OptimizationLevel {
+    None = 0,
+    Less = 1,
+    Default = 2,
+    Aggressive = 3,
+}
+
+pub struct LLVMModuleCtx {
+    functions: HashMap<RefId<ResolvedRef>, FunctionBuilder>,
+    llvm_functions: HashMap<RefId<ResolvedRef>, LLVMFunction>,
+    module: *mut LLVMModule,
+    builder: *mut LLVMBuilder,
+}
+
+impl LLVMModuleCtx {
+    pub fn generate(ir_module: &IRModule) -> Self {
+        let module = unsafe { LLVMModuleCreateWithName(c_str_ptr!(ir_module.references[&ir_module.name])) };
+        let builder = unsafe { core::LLVMCreateBuilder() };
+
+        let mut ctx = Self { 
+            functions: HashMap::new(),
+            llvm_functions: HashMap::new(),
+            module,
+            builder
+        };
+
+        ctx.generate_builtins(ir_module.references);
+
+        for func in &ir_module.functions {
+            ctx.register_function(func, ir_module.references);
+        }
+
+        for func in &ir_module.functions {
+            ctx.build_function(func)
+        }
+
+        ctx
+    }
+
+    fn generate_builtins(&mut self, references: &HashMap<RefId<ResolvedRef>, &str>) {
+        let printf_t = unsafe { LLVMInt32Type() };
+        let printf_fn_type = unsafe { LLVMFunctionType(printf_t, [ LLVMPointerType(LLVMArrayType(LLVMInt8Type(), 0), 0)].as_mut_ptr(), 1, true as LLVMBool)};
+        let printf = unsafe { LLVMAddFunction(self.module, c_str_ptr!("printf"), printf_fn_type) };
+        
+        let print_int_t = unsafe { LLVMVoidType() };
+        let print_int_fn_type = unsafe { LLVMFunctionType(print_int_t, [ LLVMInt32Type() ].as_mut_ptr(), 1, true as LLVMBool)};
+        let print_int = unsafe { LLVMAddFunction(self.module, c_str_ptr!("print_int"), print_int_fn_type) };
+        let entry = unsafe { LLVMAppendBasicBlock(print_int, c_str_ptr!("entry")) };
+        unsafe { LLVMPositionBuilderAtEnd(self.builder, entry); }
+        let value = unsafe { LLVMGetParam(print_int, 0) };
+        let fmt_str = unsafe { LLVMBuildGlobalStringPtr(self.builder, c_str_ptr!("debug: %d"), c_str_ptr!("print_int_fmt_str")) };
+        unsafe { LLVMBuildCall2(self.builder, printf_fn_type, printf, [ fmt_str, value ].as_mut_ptr(), 2, c_str_ptr!()); }
+        unsafe { LLVMBuildRetVoid(self.builder); }
+        let mut id = None;
+        for (k, v) in references {
+            if *v == "print_int" {
+                id = Some(k)
+            }
+        }
+        let id = id.unwrap();
```

Comment:
Better to use expect instead so that the intend behind the unwrap is more clear. Even better: use a non-panic option if possible